src/runtime/tests/type-test.ts

```diff
@@ -430,6 +431,173 @@ describe('types', () => {
+        assert.isTrue(varInfoMin.isAtLeastAsSpecificAs(varInfoFromAtoAbcd));
+        assert.isTrue(varInfoFromAtoAbcd.isAtLeastAsSpecificAs(varInfoMin));
+
+        // Both (undefined - {a}) and ({a} - {a, b, c, d}) are isAtLeastAsSpecificAs as each other
+        assert.isTrue(varInfoMax.isAtLeastAsSpecificAs(varInfoFromAtoAbcd));
+        assert.isTrue(varInfoFromAtoAbcd.isAtLeastAsSpecificAs(varInfoMax));
+
+        // Both (undefined - {b}) and ({a} - {a, b, c, d}) are isAtLeastAsSpecificAs as each other
+        assert.isFalse(varInfoOtherMax.isAtLeastAsSpecificAs(varInfoFromAtoAbcd));
+        assert.isFalse(varInfoFromAtoAbcd.isAtLeastAsSpecificAs(varInfoOtherMax));
+
+        // (undefined - undefined) isAtLeastAsSpecificAs ({a} - {a, b, c, d})
+        assert.isTrue(emptyVarInfo.isAtLeastAsSpecificAs(varInfoFromAtoAbcd));
+        assert.isFalse(varInfoFromAtoAbcd.isAtLeastAsSpecificAs(emptyVarInfo));
+
+        // ({a} - {a, b, c, d}) isAtLeastAsSpecificAs ({a, b} - {a, b})
+        assert.isTrue(varInfoFromAtoAbcd.isAtLeastAsSpecificAs(varInfoFromAbToAb));
+        assert.isFalse(varInfoFromAbToAb.isAtLeastAsSpecificAs(varInfoFromAtoAbcd));
+
+        // ({a} - {a, b, c, d}) isAtLeastAsSpecificAs ({a, b} - {a, b, c})
+        assert.isTrue(varInfoFromAbToAbc.isAtLeastAsSpecificAs(varInfoFromAbToAbc));
+        assert.isTrue(varInfoFromAtoAbcd.isAtLeastAsSpecificAs(varInfoFromAbToAbc));
+        assert.isFalse(varInfoFromAbToAbc.isAtLeastAsSpecificAs(varInfoFromAtoAbcd));
+
+        // None of ({a} - {a, b, c, d}) and  ({a, b} - {a, b, c, d, e}) is isAtLeastAsSpecificAs the other.
+        assert.isFalse(varInfoFromAtoAbcd.isAtLeastAsSpecificAs(varInfoFromAbtoAbcde));
+        assert.isFalse(varInfoFromAbtoAbcde.isAtLeastAsSpecificAs(varInfoFromAtoAbcd));
+
+        // None of ({a} - {a, b, c, d}) and  ({b} - {b, d}) is isAtLeastAsSpecificAs the other.
+        assert.isFalse(varInfoFromAtoAbcd.isAtLeastAsSpecificAs(varInfoFromBtoBd));
+        assert.isFalse(varInfoFromBtoBd.isAtLeastAsSpecificAs(varInfoFromAtoAbcd));
+      };
+
+      verifyIsAtLeastAsSpecificAs();
+      assert.isTrue(varInfoMin.maybeEnsureResolved());
+      verifyIsAtLeastAsSpecificAs();
+      assert.isTrue(varInfoMax.maybeEnsureResolved());
+      verifyIsAtLeastAsSpecificAs();
+      assert.isTrue(varInfoOtherMax.maybeEnsureResolved());
+      verifyIsAtLeastAsSpecificAs();
+      assert.isTrue(varInfoFromAtoAbcd.maybeEnsureResolved());
+      verifyIsAtLeastAsSpecificAs();
+      assert.isTrue(varInfoFromAbToAb.maybeEnsureResolved());
+      verifyIsAtLeastAsSpecificAs();
+      assert.isTrue(varInfoFromAbToAbc.maybeEnsureResolved());
+      verifyIsAtLeastAsSpecificAs();
+      assert.isTrue(varInfoFromAbtoAbcde.maybeEnsureResolved());
+      verifyIsAtLeastAsSpecificAs();
+      assert.isTrue(varInfoFromBtoBd.maybeEnsureResolved());
+      verifyIsAtLeastAsSpecificAs();
+    });
```

Comment:
(I might have missed them, but): Are there tests for max type variable resolution? i.e. when the last field of `TypeVariableInfo` is true.

Reply:
this test is for `isAtLeastAsSpecificAs` method of TypeVariableInfo.

TypeVariable is initialized with min and/or max bound. Then at resolution time (call to `maybeEnsureResolved`), `_resolution` is set to the selected resolved type and min and max bounds are set to null.
Long term, we want to keep the original range as well. As an incremental step towards it, as of this PR, i keep the original bounds in separate fields - `_originalCanWriteSuperset` and `_originalCanReadSubset`. For resolved variables I use the original bound values when comparing 2 variables.

I call `maybeEnsureResolved` in this test to verify that `isAtLeastAsSpecificAs` is the same for resolved and unresolved type variables.

Reply:
I have no problem with this specific test, I mean to comment about the file as a whole. Admittedly, I have limited context, so please bear with me.

Is it worth while to test ingress restricting when `TypeVariableInfo.resolveToMaxType` is set to `true` (last ctor argument of `TypeVariableInfo` is `true`)? If a recipe has a `with {*}` variable, how will that impact ingress restricting?  Such an initialization should change the result of `maybeEnsureResolved` and the `_resolution` type.

Reply:
Given your comment and these tests, maybe it's the case that how variables are resolved has no bearing on `restrictTypeRanges`.